partiql-eval/src/plan.rs

```diff
@@ -29,17 +30,49 @@ use crate::eval::expr::{
 use crate::eval::EvalPlan;
 use partiql_value::Value::Null;
 
+#[macro_export]
+macro_rules! correct_num_args_or_err {
+    ($self:expr, $args:expr, $exact_num:literal, $name:expr) => {
+        if $args.len() != $exact_num {
+            $self.errors.push(EvaluationError::IllegalState(format!(
+                "Wrong number of arguments for {}",
+                $name.to_string()
+            )));
+            return Box::new(ErrorNode::new());
+        }
+    };
+    ($self:expr, $args:expr, $min_num:literal, $max_num:literal, $name:expr) => {
+        if !($min_num..=$max_num).contains(&$args.len()) {
+            $self
+                .errors
+                .push(EvaluationError::IllegalState($name.to_string()));
+            return Box::new(ErrorNode::new());
+        }
+    };
+}
+
 #[derive(Default)]
-pub struct EvaluatorPlanner;
+pub struct EvaluatorPlanner {
+    errors: Vec<EvaluationError>,
+}
 
 impl EvaluatorPlanner {
+    pub fn new() -> EvaluatorPlanner {
+        EvaluatorPlanner { errors: vec![] }
+    }
+
     #[inline]
-    pub fn compile(&self, plan: &LogicalPlan<BindingsOp>) -> EvalPlan {
-        self.plan_eval(plan)
+    pub fn compile(&mut self, plan: &LogicalPlan<BindingsOp>) -> Result<EvalPlan, EvalErr> {
```

Comment:
Something I was less sure about was whether these logical plan to eval plan errors should be their own category (e.g. `CompileError` or `PlanningError`). Since the planning code currently resides in the `partiql-eval` crate, thought it was reasonable for now to use the same error type as the other eval errors. I'm open to creating a separate error type though.

Reply:
From the user's perspective, I think it's worthwhile to differentiate between *semantic* errors that happen while preparing a query for evaluation vs runtime errors that can only happen at evaluation time.  IMHO, it's ideal to have one `*Error` enum for every distinct step that the client application has to manually invoke.  i.e. if compilation, planning, and evaluation are three distinct steps, then one enum for each.

All such errors should have a line & column number, where applicable.

Reply:
Planning and evaluating are distinct phases. They currently live in the same crate today, but that is an implementation detail.

Reply:
Agree on the points to make planning and evaluating separate error categories. Updated the PR to have different errors:
- `PlanningError` and `PlanErr` (for multiple planning errors)
- `EvaluatorError` and `EvalErr` (for multiple evaluator errors) -- these might get adjusted further as I change the evaluation APIs to use `Result` types in an upcoming PR
---
> All such errors should have a line & column number, where applicable.

Yeah, we will get to that at some point for all applicable errors. I'll cut an issue to track, so we won't forget.